Scala 3 support

keynmol opened the pull request:
After lots of printlns I realised there was a custom scalameta-like parser for Scala 3.

All of the positions in sections had `Input.None`, so slicing wasn't working - making all tests fail.

I've added a few things on top of #443 (didn't realise it was in a fork), and now I seem to be getting 

```
[error] Failed: Total 170, Failed 54, Errors 0, Passed 107, Skipped 9
```

On the 3.0.0-M3 tests. Some of them are super easy to fix (different wording error messages in Scala3), I'll take a look at the rest which might be more complex.

tgodzik commented:
> After lots of printlns I realised there was a custom scalameta-like parser for Scala 3.
> 
> All of the positions in sections had `Input.None`, so slicing wasn't working - making all tests fail.

The positions coming from Scala 3 parser are empty. We used it before Scala 3 dialect for scalameta parser was finished. Do let me know if you need any help with Scala 3 part, that was mostly done by be :sweat_smile:

keynmol commented:
Thanks @tgodzik I sort of took a stab at it almost mechanically, now it's the hard part.

I still need to figure out @olafurpg's changes to cli module (whether they're supposed to go out?), as they introduce metaconfig's scala 2 macros and don't work for me locally. What I'm trying to figure out is if I'm missing something :D

So just to clarify:

1. We *can* now use semanticdb for latest scala
2. We are missing the dotty position <-> semanticdb position conversion, right?
3. And the custom implementation of markdown processing using dotty compiler directly can go, right?

I should've clarified those before starting with this :)

tgodzik commented:
> Thanks @tgodzik I sort of took a stab at it almost mechanically, now it's the hard part.
> 
> I still need to figure out @olafurpg's changes to cli module (whether they're supposed to go out?), as they introduce metaconfig's scala 2 macros and don't work for me locally. What I'm trying to figure out is if I'm missing something :D
> 

@olafurpg might know more.

> So just to clarify:
> 
> 1. We _can_ now use semanticdb for latest scala

Do you mean scalameta parser? Semanticdb is not used here as far as I know.

> 2. We are missing the dotty position <-> semanticdb position conversion, right?

We would most likely need to add that to use scalameta parser, but we do have those conversions for Scala 2, so this should be something similar.

> 3. And the custom implementation of markdown processing using dotty compiler directly can go, right?
> 

Yeah, if we use scalameta parser the only thing that needs to stay are separated `MarkdownCompiler` for SCala 2 and 3.

olafurpg commented:
The cli module will need to stay on 2.13 because it uses macros. 

3. Correct. That’s great news!

keynmol commented:
Ok awesome, I'll then take another stab at this over the weekend to get it into a working state hopefully, unless I hit another blocker.

keynmol commented:
Ok @olafurpg @tgodzik I had to make everything worse to make it better :D 

1. cli module cross-builds for all versions with the exception of using scala 2 macros
    
    I've separated the macros part so it can be addressed separately

2. Fixed compilation in a few places that were broken
3. Separated js modifier tests into a separate module - allowed me to delete some files but also keep it on Scala 2 (so it can be fixed separately) 
4. Updated the branch with all the latest changes
5. Fixed dependency versions (fs2 specifically)

New/remaining problems before the actual Scala 3 test fixing can start:

1. DefaultSuite:
```
==> X tests.markdown.DefaultSuite.leading-comment  0.004s munit.FailException: /home/runner/work/mdoc/mdoc/tests/unit/src/test/scala/tests/markdown/BaseMarkdownSuite.scala:89 'Only' tag is not allowed when `isCI=true`
88:  ): Unit = {
89:    test(name) {
90:      val reporter = newReporter()
    at munit.Assertions$class.fail(Assertions.scala:259)
    at munit.FunSuite.fail(FunSuite.scala:11)
    at munit.SuiteTransforms$$anonfun$munitOnlySuiteTransform$1$$anonfun$apply$1$$anonfun$apply$2.apply(SuiteTransforms.scala:58)
    at munit.SuiteTransforms$$anonfun$munitOnlySuiteTransform$1$$anonfun$apply$1$$anonfun$apply$2.apply(SuiteTransforms.scala:57)
    at munit.MUnitRunner$$anonfun$11.apply(MUnitRunner.scala:296)
    at munit.MUnitRunner$$anonfun$11.apply(MUnitRunner.scala:292)
```
2. For some reason compilation of WorksheetSuite broke on 2.11:

```
[error] /home/runner/work/mdoc/mdoc/tests/worksheets/src/test/scala/tests/worksheets/WorksheetSuite.scala:40:10: value asJava is not a member of Seq[java.nio.file.Path]
[error] possible cause: maybe a semicolon is missing before `value asJava'?
[error]         .asJava
[error]          ^
[error] one error found
```

I'm not sure what I've changed, but it's probably something extremely stupid.

keynmol commented on build.sbt:
```diff
@@ -126,6 +127,18 @@ lazy val fansiVersion = Def.setting {
   else "0.2.9"
 }
 
+lazy val fs2Version = Def.setting {
+  if (scalaVersion.value.startsWith("2.11")) "2.1.0"
+  else if (scalaVersion.value == "3.0.0-M2") "2.5.0"
+  else "2.5.3"
+}
+
+lazy val munitVersion = Def.setting {
+  if (scalaVersion.value == "3.0.0-M2")
+    "0.7.21"
+  else V.munit
+}
+
```
Previously tests weren't running on Scala 3, so we didn't have to cherry pick versions.

olafurpg commented:
The proposed plan sounds great. It’s good to separate the structural changes from fixing the tests.

keynmol commented:
Another yak that I've just realised:

CLI tests have no chance of passing on Scala 3 unless some form of args processing takes place - at the moment it's just returning default settings (because no derivation), and the CliSuite tries to process the entire docs folder.

So my questions are:

1. What's the status of metaconfig and Scala 3? I don't have the bandwidth to create Scala 3 macros for it for all the versions mdoc needs (meta api changed a lot between M2 -> M3 -> RC1)
2. Alternatively, is manual composition of settings an option? This would remove the scala 2 macros from here, making it more verbose but also potentially less effort across this many Scala 3 versions.

tgodzik commented:
@keynmol can't we leave cli on 2.13 only just as @olafurpg said?

> The cli module will need to stay on 2.13 because it uses macros.

keynmol commented:
@tgodzik 

I'm probably confused then - in Olaf's original PR, `mdoc` module `dependsOn(cli)` - which in this case means `mdoc` has to be cross-built for same versions as `cli`

Are you proposing an inverse relation? `cli` depends on `mdoc` and stays on 2.13 only? And moving CliSuite to `cli` module?

tgodzik commented:
> I'm probably confused then - in Olaf's original PR, `mdoc` module `dependsOn(cli)` - which in this case means `mdoc` has to be cross-built for same versions as `cli`
> 
> Are you proposing an inverse relation? `cli` depends on `mdoc` and stays on 2.13 only? And moving CliSuite to `cli` module?

I haven't looked closely at the PR :sweat: , but cli seems to be build only for scala 2 versions. Was that working? I think the plan was not to build it for Scala 3.  If it's impossible to reverse the relation (cli do depend on mdoc), we could put everything cli related into `scala-2` directory. I can take a look closer tomorrow and see what's going on.

tgodzik commented:
For sure we could CLI tests into scala-2 directory and run it only for Scala 2.

keynmol commented:
I hope Olaf can shed light :D 

Will take a closer look tomorrow at what interface is passed through between cli and mdoc (MainSettings?) and perhaps I just did something silly and misunderstood how the two are separated.

tgodzik commented:
> I hope Olaf can shed light :D
> 
> Will take a closer look tomorrow at what interface is passed through between cli and mdoc (MainSettings?) and perhaps I just did something silly and misunderstood how the two are separated.

Overall, I think MainSettings will just not work for Scala 3. we either need to migrate metaconfig to Scala 3 (in that case only 3.0.0-RC2 would be enough) or switch to a separate configuration library (something like [sconfig](https://github.com/ekrich/sconfig) for example). The same problem exists in scalafix https://github.com/scalacenter/scalafix/issues/1316

olafurpg commented:
I would oppose switching metaconfig with sconfig. My long-term plan is to replace metaconfig with Scalameta/moped but neither of those projects currently support Scala 3. If there’s any effort to upgrade the macros for Scala 3 then it should ideally happen in Moped, but it’s not something I expect will happen in the near future. 

The cli module can compile with 2.13 and be used from 3. This was working in my original mdoc PR making the build compile with Scala 3. There are maybe some issues with this approach if users have both sourcecode 2.13/3 on their classpath but it should be fixable.

keynmol commented:
According to [compatbility guide](https://scalacenter.github.io/scala-3-migration-guide/docs/compatibility.html#a-scala-3-module-depending-on-a-scala-213-artifact) it should work, yep.

I got thrown off because locally when I run `sbt '++3.0.0-M3 test'` (like CI does) I get:

```
...
[warn] There may be incompatibilities among your library dependencies; run 'evicted' to see detailed eviction warnings.
[info] compiling 1 Java source to /home/velvetbaldmime/projects/mdoc/mdoc-interfaces/target/scala-3.0.0-M3/classes ...
[info] compiling 1 Scala source to /home/velvetbaldmime/projects/mdoc/tests/tests/target/scala-3.0.0-M3/classes ...
[warn]
[warn]  Note: Unresolved dependencies path:
[error] sbt.librarymanagement.ResolveException: Error downloading org.scalameta:mdoc-cli_3.0.0-M3:2.2.18+26-36c4b783+20210331-2011-SNAPSH
```

And I got the same on your branch..

I've removed scala 3 from cli and pushed, will see if it behaves differently on CI.

keynmol commented on mdoc/src/main/scala-3/mdoc/internal/markdown/MarkdownCompiler.scala:
```diff
@@ -194,7 +222,8 @@ class MarkdownCompiler(
           }
-        reportMessage(vreporter, diagnostic, mpos, actualMessage)
+
+        reportMessage(vreporter, diagnostic, mpos, "\n" + actualMessage)
```
@tgodzik this might interest you as it affects worksheets. To give similar experience in markdown document, I had to prepend a newline to dotty's messages to make it look similar to what scala 2 does:

```

    """|error: yrangepos-scala3.md:3:11:
       |Found:    String
       |Required: Int
       |List[Int]("".length.toString)
       |          ^^^^^^^^^^^^^^^^^^
```

What actually comes from the compiler is `Found:    String\nRequired: Int`, unlike scala 2, which does `type mismatch:\nfound:...`

Will this new line be problematic for rendering in worksheets? (I think some of the worksheet tests got broken by this)

tgodzik replied:
It should not be problematic as it's rendered in the diagnostics of the editors. I am not sure I understand the issue here. How would it look like without the additional `\n` ?

keynmol replied:
Without the newline it'd look like this:

```
error: yrangepos-scala3.md:3:11:Found:    String
Required: Int
List[Int]("".length.toString)
          ^^^^^^^^^^^^^^^^^^ 
```

Whereas for scala 2 it looks like this:

```
error: yrangepos.md:3:11: type mismatch;
 found   : String
 required: Int
List[Int]("".length.toString)
          ^^^^^^^^^^^^^^^^^^
```

So I added the newline for consistency.

And it's the same for all the other errors from dotty. I can remove the newline, but I think it will look worse in markdown than scala 2 version :)

tgodzik replied:
I think it's fine then. Especially if it worked like that in Scala 2, then we already handled it correctly in worksheets. Anyway, thanks for letting me know!

keynmol replied:
Cool!

There are some strange errors remaining with regards to Scala 3 new syntax and mdoc generated code, but I think I'm on good track to finish all the tests this week.

If needed I will separate the PR (I started fixing tests to verify my changes to MarkdownCompiler were okay) as outlined in the original plan.

tgodzik replied:
Awesome, let me know when we should take a look at it!

keynmol commented on mdoc/src/main/scala-3/mdoc/internal/markdown/MarkdownCompiler.scala:
```diff
@@ -50,26 +53,28 @@ class MarkdownCompiler(
   }
-  private var driver = newDriver
+
+  private var context = newContext
```
After tinkering with failing warning suites, I found that Driver doesn't seem to be as important as Context

This change started making those tests pass (warnings properly surfaced)

keynmol commented on mdoc/src/main/scala-3/mdoc/internal/markdown/Instrumenter.scala:
```diff
@@ -2,105 +2,170 @@ package mdoc.internal.markdown
-import scala.meta.Name
 
-import Instrumenter.position
+class CodePrinter(ps: PrintStream, baseIndent: Int = 0, baseNest: Int = 0) {
```
The story here is that I overwrote (accidentally) the old version because it stopped being picked up.

This code printer seems to be handling all cases where correct indentation is needed, but I split it out to scala 3 sources to avoid changing the mainline scala 2 behaviour.

keynmol commented on mdoc/src/main/scala/mdoc/internal/markdown/Processor.scala:
```diff
@@ -27,13 +27,18 @@ import coursierapi.error.MultipleResolutionError
+
+  val scala =
+    if (BuildInfo.scalaBinaryVersion.startsWith("3.0"))
+      Scala3.withAllowToplevelTerms(true).withAllowToplevelStatements(true)
```
Use the scala 3 dialect

keynmol commented on tests/unit/src/test/scala/tests/markdown/ErrorSuite.scala:
```diff
@@ -102,7 +104,21 @@ class ErrorSuite extends BaseMarkdownSuite {
 
   checkError(
-    "already-defined",
+    "not-member-scala3".tag(OnlyScala3),
```
Disclosure: before I realised there's a `compat = ...`, I was splitting out tests like this. I will remove most of those.

Some tests cannot be done this way, as the test's content itself is rejected by Scala 3. For those splitting is the only option, as far as I can see.

keynmol commented on tests/unit/src/test/scala/tests/markdown/NestSuite.scala:
```diff
@@ -171,8 +185,38 @@ class NestSuite extends BaseMarkdownSuite {
        |""".stripMargin
   )
 
+  check(
```
One of the tests where it has to be split, as implicit values in Scala 3 have to have an explicit type

keynmol commented on tests/unit/src/test/scala/tests/markdown/NestSuite.scala:
```diff
@@ -231,7 +355,7 @@ class NestSuite extends BaseMarkdownSuite {
 
   checkError(
-    "implicit-nok",
+    "implicit-nok".tag(SkipScala3),
```
This is a new behaviour in Scala 3, I think, as implicits are resolved to nearest scope:

```scala
❯ cs launch scala3-repl
scala> object A {
     | implicit val x: Int = 1
     | object B {
     | implicit val y: Int = 1
     | object C{
     | implicitly[Int]
     | }
     |
     | }
     | }
// defined object A

scala> A.B.C
val res0: A.B.C.type = A$B$C$@8167f57
```

olafurpg replied:
This could be worth commenting next to the test.

keynmol commented:
@tgodzik @olafurpg @ckipp01 The magical number seems to be 15 - 15 remaining tests

They're related to file imports, scalac options, mdoc properties and 1 strange test in worksheets that I broke.

On the plus side, significant indentation is working for the generated code.

I'm running out of steam to fix the rest of the tests because I don't understand how those file imports fit in the codebase. 

Would appreciate some ideas/pointers on how to fix them, alternatively you can review what's in this PR right now (I changed a **lot** in the interactions with the compiler), and I can put skip tags on those and we can revisit them later.

keynmol commented on tests/worksheets/src/test/scala/tests/worksheets/WorksheetSuite.scala:
```diff
@@ -342,9 +342,10 @@ class WorksheetSuite extends BaseSuite {
+       |ambiguous implicit arguments: both object c1 in class App and object c2 in class App match type App.this.C of parameter c of method fn in class App
        |val xx = fn
-       |         ^^
+       |           ^
```
I thought I broke something, but it seems to be this on the latest:

```
❯ cs launch scala3-repl
scala> abstract class C:
     |   val x: Int
     | given c1 : C with
     |   val x = 1
     | given c2 : C with
     |   val x = 2
     | def fn(using c: C) = ()
     | val xx = fn
8 |val xx = fn
  |           ^
  |ambiguous implicit arguments: both object c1 and object c2 match type C of parameter c of method fn
```

tgodzik replied:
We worked around the issue, because this was causing issues for edit distance. If it's on a non-existing position we cannot match it to any real token. https://github.com/scalameta/mdoc/pull/386

tgodzik replied:
Could you open up an issue to figure it out later on?

olafurpg commented on tests/unit/src/test/scala/tests/cli/MdocPropertiesSuite.scala:
```diff
@@ -3,13 +3,12 @@ package tests.cli
     val obtained = MdocProperties.default(PathIO.workingDirectory).scalacOptions
 
-    if (BuildInfo.scalaVersion.startsWith("3.0"))
+    if (mdoc.internal.BuildInfo.scalaVersion.startsWith("3.0"))
```
FYI, I've found it very helpful to create custom MUnit tags to skip tests based on the Scala version. See https://github.com/scalameta/munit/blob/35d868403aca0e71b31b8a794472ddcbee370334/tests/shared/src/test/scala/munit/BaseSuite.scala for an example of how it's used in the MUnit codebase

olafurpg commented on build.sbt:
```diff
@@ -289,34 +316,54 @@ lazy val unit = project
     mdocJS := Some(jsdocs)
   )
-  .dependsOn(mdoc, js, testsInput, tests)
+  .dependsOn(mdoc, js, testsInput, tests, unit % "test->test")
```
Why is `% "test->test"` needed? This pattern shouldn't be necessary when you have dedicated test projects. Utility testing methods can live in `unit/src/main/scala`. I generally avoid `% "test->test"` where possible

olafurpg commented on cli/src/main/scala-3/mdoc/internal/cli/SettingsGeneric.scala:
```diff
@@ -0,0 +1,21 @@
+import metaconfig.generic.Surface
+import metaconfig.ConfEncoder
+import metaconfig.ConfDecoder
+trait SettingsGeneric {
```
Is this file used? The cli module should only compile for Scala 2.

olafurpg commented on mdoc/src/main/scala-3/mdoc/internal/markdown/Instrumenter.scala:
```diff
@@ -2,105 +2,164 @@ package mdoc.internal.markdown
-import scala.meta.Name
 
-import Instrumenter.position
+class CodePrinter(ps: PrintStream, baseIndent: Int = 0, baseNest: Int = 0) {
```
Can we move this class to a separate file named `CodePrinter.scala`.

olafurpg reviewed (changes requested):
This has been a lot of work! Thank you @keynmol for this contribution. It's fantastic to see that mdoc will soon be able to support Scala 3.

The magic import tests in `FileSuite` aren't that important to support in Scala 3. Internally, mdoc kind of has ammonite-style scripting support excluding incremental compilation and several other caveats. I never got around to fully polish that functionality into a publicly documented feature. It's totally fine to run those tests only in Scala 2.

I would recommend using MUnit tags to skip failing tests in Scala 3 and then merge this PR before the diff gets any bigger!

tgodzik reviewed (changes requested):
Great work! I think it's almost finished, I left some comments/questions.

olafurpg commented on build.sbt:
```diff
@@ -170,6 +186,22 @@ lazy val runtime = project
+    libraryDependencies ++= List(
+      "io.get-coursier" % "interface" % V.coursier,
+      "com.vladsch.flexmark" % "flexmark-all" % "0.62.2",
+      ("org.scalameta" %% "scalameta" % V.scalameta).withDottyCompat(scalaVersion.value),
```
Are you sure? Scalameta isn’t published for Scala 3

tgodzik replied:
Yeah, but we don't publish CLI for Dotty, no?
```
    scalaVersion := scala213,
    crossScalaVersions := scala2Versions,
```

keynmol replied:
On the CLI module I believe that's correct, I'll remove the compat.

keynmol replied:
This is done.

olafurpg commented on build.sbt:
```diff
@@ -211,16 +240,14 @@ lazy val mdoc = project
     )
   )
-  .dependsOn(runtime)
+  .dependsOn(runtime, cli)
```
Can you elaborate? It’s normal that the mdoc project depends on cli like in the diff. The runtime dependencies are unchanged.

tgodzik replied:
Nvm me, I think I saw something that was not there at all :fearful:

keynmol commented on mdoc/src/main/scala-3/mdoc/internal/markdown/MarkdownCompiler.scala:
```diff
@@ -148,41 +163,77 @@ class MarkdownCompiler(
+    (edit.toOriginal(start), edit.toOriginal(end - 1)) match {
       case (Right(start), Right(end)) =>
-        Position.Range(start.input, start.start, end.end).toUnslicedPosition
+          Position.Range(start.input, start.start, end.end).toUnslicedPosition
```
Basically, to make that one Worksheets test pass, I need to do 

```scala
Position.Range(start.input, start.start - 2, end.end).toUnslicedPosition
```

(and you can guess that -2 is arbitrary)

But this then breaks several other tests:

```scala
=> Diff (- obtained, + expected)
-error: yrangepos-scala3.md:3:9:
+error: yrangepos-scala3.md:3:11:
 Found:    String
 List[Int]("".length.toString)
-        ^^^^^^^^^^^^^^^^^^^^
+          ^^^^^^^^^^^^^^^^^^
```

Which is expected, given how arbitrary `2` is. You [PR](https://github.com/scalameta/mdoc/pull/386) mentions that that workaround might have issues - feels like after running all the tests on dotty something else is required.

I can keep looking for a workaround to satisfy all tests, unless you have a better suggestion?

keynmol replied:
Looking at what comes from the compiler + what's in the TokenEditDistance:

```
val [115..118) <-> val [617..620)
  [118..119) <->   [620..621)
xx [119..121) <-> xx [621..623)
  [121..122) <->   [623..624)
= [122..123) <-> = [624..625)
  [123..124) <->   [625..626)
fn [124..126) <-> fn [626..628)
\n [126..127) <-> \n [628..629)
EOF [127..127) <-> EOF [721..721)
()

 Original: dotty-ambiguous-implicit.scala:<628..628>. 
 Edited Start: 126:127, 
 Edited End: 124:126 
```

This:

```
fn [124..126) <-> fn [626..628)
```

Looks like a bug.. somewhere?

tgodzik replied:
I remember that this was not pretty, I think we can leave it as is and I can take a look in another PR? Just maybe let's track that under an issue.

tgodzik replied:
> I think we can leave it as is

keynmol replied:
I've changed the test to what it output now.

keynmol commented on cli/src/main/scala/mdoc/internal/cli/Settings.scala:
```diff
@@ -316,76 +245,25 @@ object Settings extends MetaconfigScalametaImplicits {
       usage,
       description
     ).helpMessage(width)
-  implicit val surface: Surface[Settings] =
```
I've reverted the changes

tgodzik commented on .github/workflows/ci.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
       - uses: actions/checkout@v2
       - uses: olafurpg/setup-scala@v10
-      - run: sbt docs/mdoc
+      - run: sbt '++2.12.13 docs/mdoc'
```
Isn't this the default version?

keynmol replied:
It no longer is - I had to set 2.13 to be the default for the "Scala 3 module depending on 2.13 module to work"

Here's the failure: https://github.com/scalameta/mdoc/runs/2331363147#step:4:256

`cli` module is a dependency of `mdoc`, which is a dependency of `docs` and it needs to be on 2.12 for I guess reusing the SBT plugin code

olafurpg replied:
Is this still necessary after #493? Evilplot has 2.13 releases on Sonatype

keynmol commented:
Issue opened for the token matching issue: https://github.com/scalameta/mdoc/issues/485

keynmol commented on build.sbt:
```diff
@@ -289,34 +316,54 @@ lazy val unit = project
+  .enablePlugins(BuildInfoPlugin, MdocPlugin)
+
+lazy val unitJS = project
+  .in(file("tests/unit-js"))
```
I think it's just github not showing it correctly, I just moved the files there:

https://github.com/scalameta/mdoc/pull/466/files#diff-8e6d42fca4e8e70170fa9a6048f2c04f02e34443cd25af79aac2914561ffc275

keynmol commented on build.sbt:
```diff
@@ -379,6 +432,8 @@ lazy val docs = project
       "org.scala-sbt" % "sbt" % sbtVersion.value,
       "com.cibo" %% "evilplot" % "0.6.3"
     ),
+    Compile / TaskKey[Option[File]]("bloopGenerate") := None,
```
This is from Olaf's original PR, I'll remove

keynmol commented:
> We can always merge as is and try to improve things later on as long as Scala 2 functionality remain

Yep - that's why I didn't unify things like instrumenter - to avoid breaking the Scala 2 versions, and combine them in a later issue.

keynmol commented on mdoc/src/main/scala-3/mdoc/internal/markdown/MarkdownCompiler.scala:
```diff
@@ -111,6 +116,14 @@ class MarkdownCompiler(
     report(vreporter, input, fileImports, run.runContext, edit)
   }
 
+  class CollectionReporter extends dotty.tools.dotc.reporting.Reporter {
```
Looking at it - it seems to be implemented in terms of Scala 2 models, do you mean this one? https://github.com/scalameta/mdoc/blob/main/mdoc/src/main/scala-2/mdoc/internal/markdown/FilterStoreReporter.scala (and its version specific variations)

I had to create this one because (trying to reconstruct from memory) because I was getting NPE from the [built-in one](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/reporting/StoreReporter.scala) and for warning tests I needed all the diagnostics, and by default only errors are surfaced

keynmol commented on tests/unit/src/test/scala/tests/cli/ScalacOptionsSuite.scala:
```diff
@@ -99,8 +110,9 @@ class ScalacOptionsSuite extends BaseCliSuite {
     onStdout = { out => assert(!out.contains("discarded non-Unit value")) }
   )
 
+  // TODO: fix for Scala 3
```
It does [exist](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/config/ScalaSettings.scala#L159) but it doesn't seem to do anything in REPL - is it the TPrint that prints out the type?

tgodzik replied:
For Scala 3 it's based on macros. It's not the TPrint as in case of Scala 2

olafurpg replied:
It's OK to remove the `TODO` comment. It's already explicitly documented that this test is skipped in Scala 3. If it's important to fix this test then it's better to open an issue and link to the issue from here

keynmol commented on mdoc/src/main/scala-3/mdoc/internal/markdown/MarkdownCompiler.scala:
```diff
@@ -148,41 +163,77 @@ class MarkdownCompiler(
-      position: SourcePosition
-  ): Position = {
+  def fail(edit: TokenEditDistance, input: Input, sectionPos: Position): String = {
+    val compiler = new Compiler
```
I sort of aligned it with how it's done in [Scala 2](https://github.com/scalameta/mdoc/blob/main/mdoc/src/main/scala-2/mdoc/internal/markdown/MarkdownCompiler.scala#L81-L84) - compileSources receives more arguments as it sends diagnostics to mdoc's reporter as well - not sure if making compileSources take it optionally would be an improvement

tgodzik reviewed (approved):
I think we should merge it and fix the rest of the issues in separate PRs. I don't see anything really blocking here. @olafurpg what do you think?

olafurpg commented on .github/workflows/mdoc.yml:
```diff
@@ -12,6 +12,6 @@ jobs:
       - uses: olafurpg/setup-scala@v10
       - uses: olafurpg/setup-gpg@v3
-      - run: sbt docs/docusaurusPublishGhpages
+      - run: sbt '++2.12.13 docs/docusaurusPublishGhpages'
```
Ditto, evilplot

keynmol replied:
Ah, I think I got it now.

1. We need 2.13 to be the default version in the build for the `Scala 3 module -> dependsOn Scala 2.13 module` to work
2. `2.12` is needed in docs specifically because it depends on the `plugin` which cannot be built for 2.13 (SBT)

olafurpg replied:
Hmm, why does the docs project depend on the sbt plugin? The plugin is enabled in the docs project but there shouldn’t be a project dependency there 🤔

either way, feel free to ignore this comment

olafurpg commented on mdoc/src/main/scala-3/mdoc/internal/markdown/CodePrinter.scala:
```diff
@@ -0,0 +1,47 @@
+import java.io.PrintStream
+
+class CodePrinter(ps: PrintStream, baseIndent: Int = 0, baseNest: Int = 0) {
+  private def indent = "  " * (baseIndent + nestCount)
```
What happens if the user has 3-space indent?

keynmol replied:
Pushed the test in https://github.com/scalameta/mdoc/pull/466/commits/21980022171cf20f16ca8bf106fe8c24c4605758

olafurpg replied:
Looks good. Always worth having tests like these

olafurpg reviewed (approved):
LGTM 👍 This PR represents significant work. It's a fantastic milestone to be able to support Scala 3 and Scala 2 with mdoc. Thank you @keynmol and @tgodzik for making this happen 🙏

keynmol commented:
![image](https://user-images.githubusercontent.com/1052965/117564070-0af5bf00-b0a2-11eb-894e-32bcd210d270.png)
